blog/2021-12-29-auto-setup.md

```diff
@@ -0,0 +1,121 @@
+
+<!--truncate-->
+
+![image](https://user-images.githubusercontent.com/6764957/147678999-883be1b4-4d32-4c89-84a4-00e8b701cdef.png)
+
+While the other pieces are industry standard formats with explanations available elsewhere, the Auto-Setup is the most proprietary of them, and so essential that we often recommend users just use the [temporalio/auto-setup](https://hub.docker.com/r/temporalio/auto-setup) Docker image, over our Docker image with Temporal Server itself (most beginners don’t even know this because we abstract it with docker-compose). 
+
+It’s worth: 
+
+- knowing that the `auto-setup.sh` script exists and is an important part of the Temporal deployment model,
+- understanding what it does to fill in the blanks between “it’s just a binary, any decent dev can figure it out from here right?” to production ready cluster, and
+- understanding what is optional so you can modify it when it is getting in your way, or have confidence throwing it out and writing your own.
+
+The goal of this post is ***not*** to explain every little detail (you can [read the code](https://github.com/temporalio/temporal/blob/master/docker/auto-setup.sh) yourself for that), but to highlight important pieces every self-hosted Temporal user should know as essential system knowledge for deploying Temporal Server.
+
+## Auto-Setup at a High Level
+
+Our [basic Docker entrypoint](https://github.com/temporalio/temporal/blob/master/docker/entrypoint.sh) does some bindings, but most importantly, it runs `[auto-setup.sh](https://github.com/temporalio/temporal/blob/master/docker/auto-setup.sh)` then `[start-temporal.sh](https://github.com/temporalio/temporal/blob/master/docker/start-temporal.sh)`.  
+
+`auto-setup.sh` does a few things (we will go deeper on the **bolded** parts):
+
+1. Temporal Schema Setup (if you don’t opt out):
+    1. Validates seed environment variables (trivial step)
+    2. Waits for the database to start (using `nc` for SQL databases, or [`temporal-cassandra-tool`](https://github.com/temporalio/temporal/blob/master/tools/cassandra/README.md) for Cassandra)
+    3. **Setup Temporal Schema on that database (more on this below)**
+2. Elasticsearch Schema Setup (if you opt in):
+    1. Validates seed environment variables (trivial step)
+    2. Waits for Elasticsearch to start (optionally, timed out with a `ES_SCHEMA_SETUP_TIMEOUT_IN_SECONDS` env var)
+    3. Setup Temporal’s Enhanced Visibility Schema on that Elasticsearch instance
+3. Setup Temporal Server:
+    1. Wait for Temporal Server to start (from running `[start-temporal.sh](http://start-temporal.sh)` concurrently, and checked using `tctl cluster health | grep SERVING`
+    2. **Register the `default` namespace** (unless opted out with `SKIP_DEFAULT_NAMESPACE_CREATION`)
+    3. Add basic custom search attributes (unless opted out with `SKIP_ADD_CUSTOM_SEARCH_ATTRIBUTES`)
```

Comment:
Are these the "semi-custom" attributes? If they aren't, why is it required step (there is no caveat about users needing to specify their own custom attributes.

Reply:
i dont know either, this was a surprise to find when i went thru the code. @alexshtin can perhaps offer more info here. i suspect it is just historical artifact